pkg/ctrl/doc.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package ctrl provides libraries for implementing Controllers.  Controllers are how Kubernetes APIs are implemented,
+and are central to building Operators, Workload APIs, Configuration APIs, Autoscalers, and more.
+
+Usage
+
+Create a main function that starts multiple Controllers using the ControllerManager.
+
+	pkg main
+
+	import (
+	  "flag"
+	  "log"
+
+	  "github.com/kubernetes-sigs/kubebuilder/pkg/ctrl"
+
+	  _ "pkg/controller/mycontroller"
+	)
+
+	func main() {
+	  flag.Parse()
+	  log.Fatal(ctrl.Start())
+	}
+
+Create a Controller in its own packages, and register it with the ControllerManager through the init function.
+
+	pkg mycontroller
+
+	import "github.com/kubernetes-sigs/kubebuilder/pkg/ctrl"
+
+	func init() {
+	  controller := &ctrl.Controller{Name: "pod-controller", Reconcile: ctrl.DefaultReconcileFunc}
+	  ctrl.Register(controller)
+
+	  controller.Watch(
+	    ctrl.Resource{Group: "core", Version: "v1", Enqueue: "pods"},
+	    ctrl.Enqueue{},
+	  )
+	}
+
+Controllers
```

Comment:
this section is very useful, and definitely helps me understand what's going on. It is still pretty complicated in nature (e.g. `Watch Event -> Enqueue ObjectKey(s) from Event -> Dequeue ObjectKey -> Reconcile` is not obvious ...), and thus hard to explain in plain english.

I think this overall section would be extremely useful if you could explain using a good example. As in, "OK let's say we want to build a controller that does X (scale number of pods maybe?), here are the concepts we need to know and why, and how we use them.".

The examples that you give are all mostly empty and thus hard to relate to or understand ("right, I know how to do an empty version of this, but why do I even need it since it's empty?")

Then potentially, you can use some of the other examples to show some nuance "Let's say we want to do it a different way, we could use this feature".

Reply:
I don't think I really understand what an `Event` is.

Reply:
I don't see any definition of what an `ObjectKey` is either